### PR TITLE
Generate HDF5 from TRKs 

### PR DESCRIPTION
**Pull request by frheault**
# Quick description

New script to generate a HDF5 in the same style as for connectomics from a list of TRKs.

(Tests for the script will come this week)

...

## Type of change

Check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Provide data, screenshots, command line to test (if relevant)

...

# Checklist

- [x] My code follows the style guidelines of this project (run [autopep8](https://pypi.org/project/autopep8/))
- [ ] I added relevant citations to scripts, modules and functions docstrings and descriptions
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I moved all functions from the script file (except the argparser and main) to scilpy modules
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: scripts/scil_tractogram_convert_hdf5_to_trk.py [hunk 1]
```diff
@@ -59,8 +59,9 @@ def _build_arg_parser():
                             'interest.\nEquivalent to adding any --edge_keys '
                             'node_LABEL2 or LABEL2_node.')
 
-    p.add_argument('--save_empty', metavar='labels_list', dest='labels_list',
-                   help='Save empty connections. Then, the list of possible '
+    p.add_argument('--save_empty', nargs='?', metavar='labels_list',
+                   dest='labels_list', const=True,
+                   help='Save empty connections. The list of possible '
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_hdf5_to_trk.py:64]: I would add in the description that the labels_list is now facultative.

Comment (EmmaRenauld) [reply scripts/scil_tractogram_convert_hdf5_to_trk.py:64]: Why not change the dest to `labels_file` instead. It sounds like `args.labels_list` should be a list.

**Comment by EmmaRenauld** [comment]
> (Tests for the script will come this week)

This week, hein ;)

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+"""
+Save connections as TRK to HDF5.
+
+This script is useful to convert a set of connections or bundles to a single
+HDF5 file. The HDF5 file will contain a group for each input file, with the
+streamlines stored in the specified space and origin.
+
+To make a file compatible with scil_tractogram_commit.py or
+scil_connectivity_compute_matrices.py. you will have to follow this nomenclature
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:12]: ".py." should be ".py,"

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+for the input files:
+in_dir/
+    |-- LABEL1_LABEL1.trk
+    |-- LABEL1_LABEL2.trk
+    |-- [...]
+    |-- LABEL90_LABEL90.trk
+
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:19]: Didn't read the code yet. But from a user point of view reading this doc, here's a question: what happens if one is missing? Ex, if there is not file LABEL1_LABEL2.trk. Do you infer from other files all existing labels? Do you raise a warning? Will it work nonetheless?

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+By default, ignores the empty connections. To save them, use the --save_empty.
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:20]: use the --save_empty option, or
use --save_empty.

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+from scilpy.io.utils import (add_overwrite_arg, add_verbose_arg,
+                             assert_inputs_exist,
+                             assert_outputs_exist)
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(
+        description=__doc__, formatter_class=argparse.RawTextHelpFormatter)
+    p.add_argument('in_bundles', nargs='+',
+                   help='Path of the input connections or bundles.')
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:43]: connection(s) or bundle(s).

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    p.add_argument('out_hdf5',
+                   help='Output HDF5 filename (.h5).')
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:45]: I had a question concerning this. In dwi_ml, I save my hdf5 files as .hdf5, not .h5. Any difference? Do we always accept them? Should we change the doc in all hdf5 scripts?

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Save connections as TRK to HDF5.
+
+This script is useful to convert a set of connections or bundles to a single
+HDF5 file. The HDF5 file will contain a group for each input file, with the
+streamlines stored in the specified space and origin.
+
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:10]: Maybe, something like: Most of our scripts using hdf5 files deal with any space and origin, so just choose your favorite option?

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            in_basename = os.path.splitext(os.path.basename(in_bundle))[0]
+            curr_sft = load_tractogram_with_reference(parser, args, in_bundle)
+            if len(curr_sft) == 0 and not args.save_empty:
+                logging.warning(f"Skipping {in_bundle} because it is empty")
+                continue
+
+            if not is_header_compatible(ref_sft, curr_sft):
+                parser.error(f"Header of {in_bundle} is not compatible")
+
+            # COnvert my str to the Space and Origin enums
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:87]: capital O
Convert curr_sft?

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            target_space = Space[args.stored_space.upper()]
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:88]: These two lines could be set out of the for loop.

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+                   default='vox',
+                   help='Space in which the streamlines are stored '
+                        '[%(default)s].')
+    p.add_argument('--stored_origin', choices=['nifti', 'trackvis'],
+                   default='trackvis',
+                   help='Origin in which the streamlines are stored '
+                        '[%(default)s].')
+
+    p.add_argument('--include_dps', action='store_true',
+                   help='Include the data_per_streamline the metadata.')
```
Comment (EmmaRenauld) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:57]: Note that data_per_point is never included.

File: scripts/scil_tractogram_convert_trk_to_hdf5.py (hunk 3 above)
Comment (frheault) [reply scripts/scil_tractogram_convert_trk_to_hdf5.py:19]: There is no problem with missing labels, this is why all the other scripts need a labels_list (to know what is and what is not missing).

I added: "The value of first labels should be smaller or equal to the second labels.
Connectivity scripts in scilpy only consider the upper triangular part of the
connectivity matrix."

File: scripts/scil_tractogram_convert_trk_to_hdf5.py (hunk 6 above)
Comment (frheault) [reply scripts/scil_tractogram_convert_trk_to_hdf5.py:45]: There is no difference, when I looked at stuff done in Vanderbilt and Dipy, they used .h5 so I used the same
https://figshare.com/articles/dataset/DIPY_Synb0_weights/20277522/2

File: scripts/scil_tractogram_convert_trk_to_hdf5.py (hunk 7 above)
Comment (frheault) [reply scripts/scil_tractogram_convert_trk_to_hdf5.py:10]: All of my scripts for connectivity assume vox/corner (to avoid as much computation on 100k connections). 

So I added: (keep the default if you are going to use the connectivity scripts in scilpy).

File: scripts/scil_tractogram_assign_uniform_color.py [hunk 11]
```diff
@@ -102,6 +102,7 @@ def main():
                                 '.trk file.')
             out_filenames.append('{}{}{}'
                                  .format(base, args.out_suffix, '.trk'))
+    print(out_filenames)
```
Comment (arnaudbore) [inline scripts/scil_tractogram_assign_uniform_color.py:105]:
```suggestion
```

File: scripts/scil_tractogram_convert_trk_to_hdf5.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+                   help='Path of the input connection(s) or bundle(s).')
+    p.add_argument('out_hdf5',
+                   help='Output HDF5 filename (.h5).')
+
+    p.add_argument('--stored_space', choices=['rasmm', 'voxmm', 'vox'],
+                   default='vox',
+                   help='Space in which the streamlines are stored '
+                        '[%(default)s].')
+    p.add_argument('--stored_origin', choices=['nifti', 'trackvis'],
+                   default='trackvis',
```
Comment (arnaudbore) [inline scripts/scil_tractogram_convert_trk_to_hdf5.py:57]: I don't understand this. "Origin in which streamlines are stored" and why trackvis is the default.

File: scripts/tests/test_tractogram_assign_uniform_color.py [hunk 13]
```diff
@@ -37,7 +37,11 @@ def test_execution_dict(script_runner, monkeypatch):
     json_file = 'my_json_dict.json'
     with open(json_file, "w+") as f:
         json.dump(my_dict, f)
-
+    
+    filename = os.path.splitext(in_bundle)[0]
+    filename = f'{filename}_colored.trk'
+    if os.path.isfile(filename):
+        os.remove(filename)
```
Comment (arnaudbore) [inline scripts/tests/test_tractogram_assign_uniform_color.py:44]: why did you add this... ?

**Review by arnaudbore — COMMENTED** [review]
A couple of questions, almost GTG

File: scripts/tests/test_tractogram_assign_uniform_color.py (hunk 13 above)
Comment (frheault) [reply scripts/tests/test_tractogram_assign_uniform_color.py:44]: Because it was crashing locally before.

File: scripts/scil_tractogram_convert_trk_to_hdf5.py (hunk 12 above)
Comment (frheault) [reply scripts/scil_tractogram_convert_trk_to_hdf5.py:57]: I changed to 'Voxel origin convention' and the default is "trackvis" because we mostly work with TRK and that's their convention.

(The --space and --origin default are made to work with scil_tractogram_convert_hdf5_to_trk.py so the default of this script is what is expected by scil_tractogram_convert_hdf5_to_trk.py)

File: scripts/tests/test_tractogram_assign_uniform_color.py (hunk 13 above)
Comment (arnaudbore) [reply scripts/tests/test_tractogram_assign_uniform_color.py:44]: the -f in the test should be enough instead of this.

**Review by arnaudbore — COMMENTED** [review]
GTG
